[receiver/kafka] Add rack_id configuration option to Kafka receiver

piotrbelina opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

Adds support for rack-aware replica selection in the Kafka receiver by introducing a new `rack_id` configuration parameter. When set and brokers are configured with a rack-aware replica selector, the client will prefer fetching from the closest replica, potentially reducing latency and improving performance.

Changes:
- Add Rack field to ClientConfig with mapstructure:"rack_id"
- Wire rack_id to Sarama via saramaConfig.RackID
- Wire rack_id to franz-go via kgo.Rack() option
- Update kafkareceiver README.md documentation

The configuration defaults to empty string (disabled) to maintain backward compatibility. This maps to Kafka's standard "client.rack" setting while using a more Go-idiomatic configuration key name.


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Implements #42313 

#### Testing

```yaml
receivers:
  kafka:
    brokers: ["localhost:9092"]
    rack_id: "rack1"

processors:
  batch: {}

exporters:
  debug: {}

service:
  pipelines:
    traces:
      receivers: [kafka]
      processors: [batch]
      exporters: [debug]
    metrics:
      receivers: [kafka]
      processors: [batch]
      exporters: [debug]
    logs:
      receivers: [kafka]
      processors: [batch]
      exporters: [debug]

```

<!--Describe the documentation added.-->
#### Documentation
updated README.md
<!--Please delete paragraphs that you did not use before submitting.-->

piotrbelina committed:
[receiver/kafka] Add rack_id configuration option to Kafka receiver

Adds support for rack-aware replica selection in the Kafka receiver
by introducing a new `rack_id` configuration parameter. When set and
brokers are configured with a rack-aware replica selector, the client
will prefer fetching from the closest replica, potentially reducing
latency and improving performance.

Changes:
- Add Rack field to ClientConfig with mapstructure:"rack_id"
- Wire rack_id to Sarama via saramaConfig.RackID
- Wire rack_id to franz-go via kgo.Rack() option
- Update kafkareceiver README.md documentation
- Update sample configuration with rack_id example

The configuration defaults to empty string (disabled) to maintain
backward compatibility. This maps to Kafka's standard "client.rack"
setting while using a more Go-idiomatic configuration key name.

Signed-off-by: Piotr Belina <piotr@piotrbelina.com>

axw commented on pkg/kafka/configkafka/config.go:
```diff
@@ -48,6 +48,11 @@ type ClientConfig struct {
 	// Metadata holds metadata-related configuration for producers and consumers.
 	Metadata MetadataConfig `mapstructure:"metadata"`
+
+	// Rack provides the rack identifier for this client to enable rack-aware
+	// replica selection when supported by the brokers. This maps to Kafka's
+	// standard "client.rack" setting. By default, this is empty.
+	Rack string `mapstructure:"rack_id"`
```
```suggestion
	// RackID provides the rack identifier for this client to enable rack-aware
	// replica selection when supported by the brokers. This maps to Kafka's
	// standard "client.rack" setting. By default, this is empty.
	RackID string `mapstructure:"rack_id"`
```

Let's keep the field name and the JSON/YAML tag name aligned

piotrbelina replied:
sure. I added your suggestion + changelog

axw reviewed (commented):
Thanks @piotrbelina! Looks good, just one minor request to rename the field name.
Please also create a changelog entry - see https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

piotrbelina committed:
Apply code review suggestions and add changelog

- Rename Rack field to RackID to align with mapstructure tag
- Update all references to use RackID instead of Rack
- Add changelog entry for rack_id configuration enhancement

piotrbelina committed:
Fix tests for RackID field rename

piotrbelina committed:
Merge branch 'main' into feat/add-kafka-receiver-rack-id

axw reviewed (approved):
Thank you :)
